pkg/daemon/daemon.go

```diff
@@ -791,27 +791,60 @@ func (dn *Daemon) getNodeMachinePool() error {
 	return fmt.Errorf("getNodeMachinePool(): Failed to find the MCP of the node")
 }
 
+func (dn *Daemon) getDrainLock(ctx context.Context, done chan bool) {
+	var err error
+
+	lock := &resourcelock.LeaseLock{
+		LeaseMeta: metav1.ObjectMeta{
+			Name:      "config-daemon-draining-lock",
+			Namespace: namespace,
+		},
+		Client: dn.kubeClient.CoordinationV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity: dn.name,
+		},
+	}
+
+	// start the leader election
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+		Lock:            lock,
+		ReleaseOnCancel: true,
+		LeaseDuration:   5 * time.Second,
+		RenewDeadline:   3 * time.Second,
+		RetryPeriod:     1 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				glog.V(2).Info("drainNode(): started leading")
+				for {
+					time.Sleep(1 * time.Second)
+					if dn.drainable {
+						glog.V(2).Info("drainNode(): no other node is draining")
+						err = dn.annotateNode(dn.name, annoDraining)
+						if err != nil {
+							glog.Errorf("drainNode(): Failed to annotate node: %v", err)
+							continue
+						}
+						done <- true
+						return
+					}
+				}
+			},
+			OnStoppedLeading: func() {
+				glog.V(2).Info("drainNode(): stopped leading")
+			},
+		},
+	})
+}
+
 func (dn *Daemon) drainNode(name string) error {
 	glog.Info("drainNode(): Update prepared")
 	var err error
-
 	ctx, cancel := context.WithCancel(context.TODO())
 	defer cancel()
-	// wait a random time to avoid all the nodes drain at the same time
-	time.Sleep(wait.Jitter(3*time.Second, 3))
-	wait.JitterUntil(func() {
-		if !dn.drainable {
-			glog.V(2).Info("drainNode(): other node is draining")
-			return
-		}
-		glog.V(2).Info("drainNode(): no other node is draining")
-		err = dn.annotateNode(dn.name, annoDraining)
-		if err != nil {
-			glog.Errorf("drainNode(): Failed to annotate node: %v", err)
-			return
-		}
-		cancel()
-	}, 3*time.Second, 3, true, ctx.Done())
+
+	done := make(chan bool)
+	go dn.getDrainLock(ctx, done)
```

Comment:
Can we run `dn.getDrainLock` in foreground and not use `done` chan? 
I assume `leaderelection.RunOrDie` is a loop w/o timeout, is it true?

Reply:
The `leaderelection.RunOrDie` runs forever unless the context is canceled. I didn't run it in the foreground, because I want it keeps leading until the node finishes draining. So the rest nodes can only start the election after the prior one finishes draining if there is no reboot required.